pkg/mvtxdaemon/server/service/health_status.go

```diff
@@ -0,0 +1,226 @@
+	Code string `json:"code"`
+}
+
+// newDataHealthResponse is used to create a new dataHealthResponse object
+func newDataHealthResponse(status string, message string, code string) *dataHealthResponse {
+	return &dataHealthResponse{
+		Status:  status,
+		Message: message,
+		Code:    code,
+	}
+}
+
+// defaultDataHealthResponse is the default response returned by the data health check API
+var defaultDataHealthResponse = newDataHealthResponse(
+	v1alpha1.MonoVertexStatusUnknown,
+	"MonoVertex data flow is in an unknown state",
+	"D4")
+
+// HealthChecker is the struct type for health checker.
+type HealthChecker struct {
+	// field for the health status.
+	currentDataStatus *dataHealthResponse
+	// Spec of the monoVertex getting processed
+	monoVertex *v1alpha1.MonoVertex
+	// Mutex for state information
+	statusLock *sync.RWMutex
+}
+
+// NewHealthChecker creates a new object HealthChecker struct type.
+func NewHealthChecker(monoVertex *v1alpha1.MonoVertex) *HealthChecker {
+	// Return a new HealthChecker struct instance.
+	return &HealthChecker{
+		currentDataStatus: defaultDataHealthResponse,
+		monoVertex:        monoVertex,
+		statusLock:        &sync.RWMutex{},
+	}
+}
+
+// monoVtxState is a struct which contains the name and data health state
+// of a MonoVertex
+type monoVtxState struct {
+	// Name is the name of the vertex
+	Name string `json:"name"`
+	// State is the state of the vertex
+	State string `json:"state"`
+}
+
+// newMonoVtxState is used to create a new monoVtxState object
+func newMonoVtxState(name string, state string) *monoVtxState {
+	return &monoVtxState{
+		Name:  name,
+		State: state,
+	}
+}
+
+// getCurrentHealth returns the current health status of the MonoVertex.
+// It is thread safe to ensure concurrent access.
+func (hc *HealthChecker) getCurrentHealth() *dataHealthResponse {
+	// Lock the statusLock to ensure thread safety.
+	hc.statusLock.RLock()
+	defer hc.statusLock.RUnlock()
+	// Return the health status.
+	return hc.currentDataStatus
+}
+
+// setCurrentHealth sets the current health status of the MonoVertex.
+// It is thread safe to ensure concurrent access.
+func (hc *HealthChecker) setCurrentHealth(status *dataHealthResponse) {
+	// Lock the statusLock to ensure thread safety.
+	hc.statusLock.Lock()
+	defer hc.statusLock.Unlock()
+	// Set the health status.
+	hc.currentDataStatus = status
+}
+
+// getMonoVertexDataCriticality is used to provide the data criticality of the MonoVertex
+// They can be of the following types:
+// 1. Healthy: The MonoVertex is working as expected
+// 2. Warning: The MonoVertex is working but there could be a lag in the data movement
+// 3. Critical: The MonoVertex is not working as expected
+// We need to check the following things to determine the data criticality of the MonoVertex:
+// At any given instant of time what is the desired number of replicas required by the MonoVertex
+// to clear out the backlog in the target state time.
```

Comment:
What about the case when pods are continuously being restarted? or what if the TPS is low and pending and is growing very slowly?

Reply:
It looks like cases like pod restarting will be caught by resource health.